Internalize mount.Interface.IsMountPointMatch

codenrhoden opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
IsMountPointMatch() had no callers outside of the mount package, and has
internal implementation details. This patch makes it no longer be
public.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**:
As part of https://github.com/kubernetes/utils/pull/100, it was asked why this method was needed rather than just being a simple `mp.Path == dir`. After looking closer, this method does not have an external callers and does not need to be public.

/assign @msau42

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.
Please use the following format for linking documentation or pass the
section below:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

msau42 commented on pkg/util/mount/types/types.go:
```diff
@@ -30,6 +30,8 @@ type Interface interface {
 	// to be consistent.
 	List() ([]MountPoint, error)
 	// IsMountPointMatch determines if the mountpoint matches the dir.
+	// On Linux systems, this method handles the case where the mountpoint has
```
Do we want to specifically call out implementation specifics in the interface comment? Maybe something more generic like "In some environments, the directory could be renamed if the mount point is stale".

Or, from a completely different angle, does this have to be exposed as an interface method? Or can we completely hide it in the implementation?

codenrhoden replied:
yeah, there is where I need help. :) One of the review comments was "why is this needed?" and another was "don't mention implementation details." But the reason it's needed is because the implementation is different on Linux. So we could just call out what's different in the Linux implementations (`mount_linux.go`, `nsenter`), but that leaves the question of what to say here...

But you are right... from a different angle, this doesn't appear needed in the interface. It's only called in one place: https://github.com/kubernetes/kubernetes/blob/d951d19ec4d8ec829e83d146200ea91b664b6fa6/pkg/util/mount/mount.go#L252

It's completely internal. We need different implementations for Windows/Linux, but that doesn't mean it has to be in the interface. I'll explore that.

codenrhoden commented:
@msau42  See what you think of this approach.

I also changed my mind on basing this PR on top of another PR. I'll just deal with the merge conflicts as they happen. I'd rather make obvious progress as we can...

codenrhoden commented:
/hold cancel

codenrhoden commented:
/remove-area kubelet
/remove-area test
/remove-sig apps
/remove-sig node
/remove-sig testing

msau42 commented on pkg/util/mount/mount_helper_common.go:
```diff
@@ -97,3 +98,18 @@ func PathExists(path string) (bool, error) {
+	if mp.Path == dir {
+		return true
+	}
+	if runtime.GOOS == "linux" {
```
This is a bit odd. My other thought was making IsNotMountPoint part of the interface.

codenrhoden replied:
Sure, let me know how you want me to proceed...

Before I went this direction, I did a quick search of the codebase, and saw a lot of places in `pkg/kubelet` and `pkg/volume` that were doing switches on `runtime.GOOS`, so I kind of ran with that precedent. :) It made for more compact code that could just be one common internal function.

msau42 replied:
Let's get @saad-ali 's thoughts on this.  I think it's a little odd because we already have os-specific implementations of the mounter.

hogepodge replied:
Are there instances where this code wouldn't be run on Linux, and would that be a bug? I would prefer to err on the side of safety if the mounter code somehow winds up as the default code path.

saad-ali replied:
I agree with the higher level goal here which is remove `IsMountPointMatch(...)` from the Mount interface since it is an implementation detail for Mount.

There are already `mount_helper_windows.go` and `mount_helper_unix.go` files, why not make `isMountPointMatch(...)` arch specific, and put a definition in each of those files?

codenrhoden replied:
I had tried that approach originally, but was running into issues because I thought I needed an  `unsupported` implementation along with Linux and Windows. I had failed to notice that the `mount_helper_unix.go` file is actually `+build !windows`, meaning that it builds on Darwin as well. Therefore anything in there is actually for *both* Linux and Darwin. Before I was getting a redeclaration error.

I've fixed this. Should be good to review again.

saad-ali replied:
> I had failed to notice that the mount_helper_unix.go file is actually +build !windows, meaning that it builds on Darwin as well. Therefore anything in there is actually

Unrelated to this PR, but why are we doing `+build !windows` rather then `+build linux`?

msau42 commented:
/assign @saad-ali

codenrhoden commented:
/test pull-kubernetes-bazel-test

codenrhoden commented:
/test pull-kubernetes-e2e-gce-100-performance

msau42 commented:
/lgtm

msau42 commented:
/priority important-soon

hogepodge commented:
/lgtm

codenrhoden commented:
/assign @jsafrane

codenrhoden committed:
Internalize mount.Interface.IsMountPointMatch

IsMountPointMatch() had no callers outside of the mount package, and has
internal implementation details. This patch makes it no longer be
public.

codenrhoden commented:
/hold
I'm holding this because if it got final approval, it would potentially go in before https://github.com/kubernetes/kubernetes/pull/81794, and I *really* want to avoid rebasing that one if I can. Rebasing this one is a snap.

saad-ali reviewed (approved):
/lgtm
/approve

codenrhoden commented:
/hold cancel

codenrhoden commented:
/hold

I thought the needs-rebase label would be on this automatically and it wouldn’t get into the merge pool. I was wrong. Can’t rebase until tomorrow morning, will remove hold after that. Apologies for any spam.
